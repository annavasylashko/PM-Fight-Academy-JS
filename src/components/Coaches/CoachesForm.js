import React from 'react';

import Input from '../Input/Input';
import CRUDButtons from '../buttons/CRUDButtons/CRUDButtons';

const CoachesForm = (props) => {
  return (
    <form data-testid="coach-form" action="" onSubmit={props.submitHandler}>
      <Input
        onBlur={props.blurHandler}
        onChange={props.inputHandler}
        type="text"
        name={props.inputs.firstName.name}
        defaultValue={props.inputs.firstName.value}
        error={props.inputs.firstName?.error}
        testid="form-firstname"
      />
      <Input
        onBlur={props.blurHandler}
        onChange={props.inputHandler}
        type="text"
        name={props.inputs.lastName.name}
        defaultValue={props.inputs.lastName.value}
        error={props.inputs.lastName?.error}
        testid="form-lastname"
      />
      <Input
        onBlur={props.blurHandler}
        onChange={props.inputHandler}
        type="text"
        name={props.inputs.description.name}
        defaultValue={props.inputs.description.value}
        error={props.inputs.description?.error}
        testid="form-description"
      />
      <Input
        onBlur={props.blurHandler}
        onChange={props.inputHandler}
        type="date"
        name={props.inputs.dateBirth.name}
        defaultValue={props.inputs.dateBirth.value}
        error={props.inputs.dateBirth?.error}
        testid="form-birthdate"
      />
      <Input
        onBlur={props.blurHandler}
        onChange={props.inputHandler}
        type="tel"
        name={props.inputs.phoneNumber.name}
        defaultValue={props.inputs.phoneNumber.value}
        error={props.inputs.phoneNumber?.error}
        testid="form-phone"
      />
      <CRUDButtons
        updateHandler={props.updateHandler}
        deleteHandler={props.deleteHandler}
        isCreateButtonVisible={props.isCreateButtonVisible}
        isCreateButtonDisabled={props.isCreateButtonDisabled}
        isUpdateButtonVisible={props.isUpdateButtonVisible}
        isUpdateButtonDisabling={props.isUpdateButtonDisabling}
        isDeleteButtonVisible={props.isDeleteButtonVisible}
        isDeleteButtonDisabled={props.isDeleteButtonDisabled}
      />
    </form>
  );
};

export default CoachesForm;
