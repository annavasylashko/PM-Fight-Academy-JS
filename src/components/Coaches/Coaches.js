import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  addCoach,
  getCoaches,
  updateCoach,
  deleteCoach,
  resetCoach,
} from "../../redux/actions";
import styles from "../MainStyles/mainStyles.module.css";
import List from "../List/List";
import AddButton from "../buttons/AddButton/AddButton";
import CoachesForm from "./CoachesForm";

const Coaches = () => {
  const dispatch = useDispatch();

  const coaches = useSelector((state) => state.coach.data);
  const [coach, setUpdateCoach] = useState({});

  const [isCreateButtonDisabled, setCreateButtonDisabling] = useState(true);
  const [isDeleteButtonDisabled, setDeleteButtonDisabling] = useState(true);
  const [isUpdateButtonDisabling, setUpdateButtonDisabling] = useState(true);

  const [isCreateButtonVisible, setCreateButtonVisibility] = useState(true);
  const [isDeleteButtonVisible, setDeleteButtonVisibility] = useState(false);
  const [isUpdateButtonVisible, setUpdateButtonVisibility] = useState(false);

  const initialInputs = {
    firstName: {
      name: "firstName",
      value: "",
      error: "",
    },
    lastName: {
      name: "lastName",
      value: "",
      error: "",
    },
    birthDate: {
      name: "birthDate",
      value: "",
      error: "",
    },
    description: {
      name: "description",
      value: "",
      error: "",
    },
    phoneNumber: {
      name: "phoneNumber",
      value: "",
      error: "",
    },
  };

  const [inputs, setInputs] = useState({ ...initialInputs });

  useEffect(() => {
    dispatch(getCoaches());
    return () => dispatch(resetCoach());
  }, []);

  const initialFormState = () => {
    setCreateButtonDisabling(true);
    setCreateButtonVisibility(true);
    setDeleteButtonVisibility(false);
    setUpdateButtonVisibility(false);
    setInputs((state) => ({ ...initialInputs }));
  };

  const inputHandler = (e) => {
    setInputs((state) => {
      let new_state = { ...state };
      new_state[e.target.name].value = e.target.value;
      new_state[e.target.name].error = "";
      return new_state;
    });
    if (
      e.target.value &&
      [
        inputs.phoneNumber.value,
        inputs.lastName.value,
        inputs.firstName.value,
        inputs.description.value,
      ].every((i) => i.length > 0)
    ) {
      setCreateButtonDisabling(false);
      setUpdateButtonDisabling(false);
    }
  };

  const blurHandler = (e) => {
    if (e.target.value.length < 1) {
      setInputs((state) => {
        let new_state = { ...state };
        new_state[e.target.name].error = "Empty field";
        return new_state;
      });
      setCreateButtonDisabling(true);
    }
  };

  const submitHandler = (e) => {
    e.preventDefault();
    dispatch(
      addCoach({
        id: coaches.sort((a, b) => b.id - a.id)[0].id + 1,
        firstName: inputs.firstName.value,
        lastName: inputs.lastName.value,
        birthDate: inputs.birthDate.value,
        description: inputs.description.value,
        phoneNumber: inputs.phoneNumber.value,
      })
    );
    setInputs((state) => ({ ...initialInputs }));
  };

  const updateHandler = () => {
    dispatch(
      updateCoach({
        firstName: inputs.firstName.value,
        lastName: inputs.lastName.value,
        birthDate: inputs.birthDate.value,
        description: inputs.description.value,
        phoneNumber: inputs.phoneNumber.value,
        id: coach.id,
      })
    );
    initialFormState();
  };

  const deleteHandler = () => {
    dispatch(deleteCoach(coach.id));
    initialFormState();
  };

  const addForm = (e) => {
    initialFormState();
  };

  const setCoach = (coachId) => {
    return (e) => {
      setCreateButtonVisibility(false);
      setDeleteButtonVisibility(true);
      setUpdateButtonVisibility(true);
      setDeleteButtonDisabling(false);

      const coach = coaches.find((coach) => coach.id === coachId);
      setUpdateCoach(coach);

      const serviceInput = {
        firstName: {
          name: "firstName",
          value: coach.firstName,
          error: "",
        },
        lastName: {
          name: "lastName",
          value: coach.lastName,
          error: "",
        },
        birthDate: {
          name: "birthDate",
          value: coach.birthDate,
          error: "",
        },
        description: {
          name: "description",
          value: coach.description,
          error: "",
        },
        phoneNumber: {
          name: "phoneNumber",
          value: coach.phoneNumber,
          error: "",
        },
      };
      setInputs((state) => ({ ...serviceInput }));
    };
  };

  const formatCoaches = () =>
    coaches?.map((coach) => (
      <p key={coach.id} onClick={setCoach(coach.id)}>
        {coach.firstName} {""}
        {coach.lastName}
      </p>
    ));

  return (
    <div className={styles.container}>
      <div className={styles.leftContainer}>
        <AddButton addForm={addForm} />
        <List items={formatCoaches()} />
      </div>

      <div className={styles.rightContainer}>
        <CoachesForm
          inputs={inputs}
          blurHandler={blurHandler}
          inputHandler={inputHandler}
          submitHandler={submitHandler}
          updateHandler={updateHandler}
          deleteHandler={deleteHandler}
          isCreateButtonVisible={isCreateButtonVisible}
          isCreateButtonDisabled={isCreateButtonDisabled}
          isUpdateButtonVisible={isUpdateButtonVisible}
          isUpdateButtonDisabling={isUpdateButtonDisabling}
          isDeleteButtonVisible={isDeleteButtonVisible}
          isDeleteButtonDisabled={isDeleteButtonDisabled}
        />
      </div>
    </div>
  );
};

export default Coaches;
