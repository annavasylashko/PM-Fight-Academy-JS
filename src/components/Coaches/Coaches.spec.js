import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Coaches from "./Coaches";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: () => [
    {
      id: 1,
      firstName: "test",
      lastName: "user",
      dateBirth: "2001-01-01",
      description: "123",
      phoneNumber: "123",
    },
  ],
}));

jest.mock("../../redux/actions", () => ({
  ...jest.requireActual("../../redux/actions"),
  addCoach: () => ({ type: "some/type" }),
}));

describe("Coaches", () => {
  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <Coaches />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("blurHandler should trigger on input and on blur", async () => {
    const { getByTestId, findByTestId } = render(
      <Provider store={store}>
        <Coaches />
      </Provider>
    );

    const firstname = getByTestId("form-firstname");
    const lastname = getByTestId("form-lastname");
    const description = getByTestId("form-description");
    const birthdate = getByTestId("form-birthdate");
    const phone = getByTestId("form-phone");

    const value = {
      target: { value: "2020-05-24" },
    };
    const emptyValue = {
      target: { value: "" },
    };

    const submitButton = getByTestId("crud-button-submit");
    const addButton = getByTestId("add-button");
    const coachForm = getByTestId("coach-form");

    expect(submitButton).toBeDisabled();

    // fill the form
    fireEvent.change(firstname, value);
    fireEvent.change(lastname, value);
    fireEvent.change(description, value);
    fireEvent.change(birthdate, value);
    fireEvent.change(phone, value);

    expect(submitButton).not.toBeDisabled();

    // check blur
    fireEvent.change(phone, emptyValue);
    fireEvent.blur(phone);

    expect(submitButton).toBeDisabled();

    // fill the form
    fireEvent.change(phone, value);

    expect(submitButton).not.toBeDisabled();

    // check add button
    fireEvent.click(addButton);

    expect(submitButton).toBeDisabled();

    // check submit
    fireEvent.change(firstname, value);
    fireEvent.change(lastname, value);
    fireEvent.change(description, value);
    fireEvent.change(birthdate, value);
    fireEvent.change(phone, value);
    expect(submitButton).not.toBeDisabled();

    fireEvent.submit(coachForm);
    expect(submitButton).toBeDisabled();

    // check update
    const listItem = await findByTestId("list-item");
    expect(listItem).toBeInTheDocument();

    await fireEvent.click(listItem);

    const updateButton = getByTestId("crud-button-update");
    const deleteButton = getByTestId("crud-button-delete");

    expect(updateButton).not.toBeDisabled();
    expect(deleteButton).not.toBeDisabled();

    await fireEvent.click(updateButton);
    expect(submitButton).toBeDisabled();
  });
});
