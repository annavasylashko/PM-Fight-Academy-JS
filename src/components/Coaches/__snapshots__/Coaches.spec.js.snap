// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coaches should render properly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="leftContainer"
    >
      <button
        class="addBtn"
        data-testid="add-button"
      >
        <p
          class="addBtnText"
        >
          Add
        </p>
        <div
          class="addBtnPlus"
        >
          +
        </div>
      </button>
      <div
        class="resultList"
      >
        <p
          data-testid="list-item"
        >
          test
           
          
          user
        </p>
      </div>
    </div>
    <div
      class="rightContainer"
    >
      <form
        action=""
        data-testid="coach-form"
      >
        <div
          class="omrs-input-group"
        >
          <label
            class="omrs-input-underlined"
          >
            <input
              data-testid="form-firstname"
              name="firstName"
              type="text"
              value=""
            />
            <span
              class="omrs-input-label"
            >
              firstName
            </span>
            <span
              class="omrs-input-helper"
            />
          </label>
        </div>
        <div
          class="omrs-input-group"
        >
          <label
            class="omrs-input-underlined"
          >
            <input
              data-testid="form-lastname"
              name="lastName"
              type="text"
              value=""
            />
            <span
              class="omrs-input-label"
            >
              lastName
            </span>
            <span
              class="omrs-input-helper"
            />
          </label>
        </div>
        <div
          class="omrs-input-group"
        >
          <label
            class="omrs-input-underlined"
          >
            <input
              data-testid="form-description"
              name="description"
              type="text"
              value=""
            />
            <span
              class="omrs-input-label"
            >
              description
            </span>
            <span
              class="omrs-input-helper"
            />
          </label>
        </div>
        <div
          class="omrs-input-group"
        >
          <label
            class="omrs-input-underlined"
          >
            <input
              data-testid="form-birthdate"
              name="dateBirth"
              type="date"
              value=""
            />
            <span
              class="omrs-input-label"
            >
              dateBirth
            </span>
            <span
              class="omrs-input-helper"
            />
          </label>
        </div>
        <div
          class="omrs-input-group"
        >
          <label
            class="omrs-input-underlined"
          >
            <input
              data-testid="form-phone"
              name="phoneNumber"
              type="tel"
              value=""
            />
            <span
              class="omrs-input-label"
            >
              phoneNumber
            </span>
            <span
              class="omrs-input-helper"
            />
          </label>
        </div>
        <div
          class="btnContainer"
        >
          <button
            class="addBtn"
            data-testid="crud-button-submit"
            disabled=""
            type="submit"
          >
            <p
              class="addBtnText"
            >
              Done
            </p>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
